### PR TITLE
ZCash transactions

### PR DESCRIPTION
**Pull request by cypt4**
Resolvers https://github.com/brave/brave-browser/issues/32304

Adds a set of transaction-related classes for ZCash (without signing step yet).

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/zcash/zcash_transaction.cc [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+  result.utxo_address = address;
+  // result.utxo_outpoint.txid = utxo.txid();
+  result.utxo_outpoint.index = utxo.index();
+  result.utxo_value = utxo.valuezat();
+  return result;
+}
+
+uint32_t ZCashTransaction::TxInput::n_sequence() const {
+  // Fixed value by now.
+  // https://github.com/ZCash/ZCash/blob/v24.0/src/wallet/spend.cpp#L945
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_transaction.cc:226]: not valid link

File: components/brave_wallet/browser/brave_wallet_utils.cc [hunk 2]
```diff
@@ -1508,6 +1540,10 @@ std::string GetKnownNetworkId(mojom::CoinType coin,
   if (coin == mojom::CoinType::BTC) {
     return GetKnownBtcNetworkId(chain_id);
   }
+  if (coin == mojom::CoinType::ZEC) {
+    return GetKnownZecNetworkId(chain_id);
+  }
+  NOTREACHED();
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_utils.cc:1546]: nit: 
`NOTREACHED() << coin;`

File: components/brave_wallet/common/brave_wallet.mojom [hunk 3]
```diff
@@ -1393,11 +1394,22 @@ struct BitcoinAccountInfo {
...
+
 struct ZCashBalance {
   uint64 total_balance;
   map<string, uint64> balances;  // address -> balance of that address(sum of all related utxos)
 };
 
+struct ZCashAddress {
+  string address_string;
+  ZCashKeyId key_id;
+};
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet.mojom:1411]: Please place these just before ZCashWalletService definition

File: components/brave_wallet/browser/zcash/zcash_wallet_service.cc [hunk 4]
```diff
@@ -57,22 +261,53 @@ void ZCashWalletService::Bind(
...
+  auto id = mojom::ZCashKeyId::New(account_id->bitcoin_account_index, 0, 0);
+  auto str_addr = keyring_service_->GetZCashAddress(*account_id, *id);
+  if (!str_addr) {
+    std::move(callback).Run(nullptr, "Failed to retreive new receiver address");
+  }
+  // TODO(cypt4): Return unused receiver address
+  std::move(callback).Run(mojom::ZCashAddress::New(*str_addr, std::move(id)),
+                          absl::nullopt);
+}
+
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_wallet_service.cc:293]: Should that be BraveWalletService.GenerateReceiveAddress?

File: components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      <TransactionFiatAmountBig>
+        {
+          new Amount(transactionDetails.fiatValue).formatAsFiat(defaultCurrencies.fiat)
+        }
+      </TransactionFiatAmountBig>
+
+      <TabRow>
+        <PanelTab
+          isSelected={selectedTab === 'transaction'}
+          onSubmit={onSelectTab('transaction')}
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx:158]: Can inline this action

```suggestion
          onSubmit={() => setSelectedTab('transaction')}
```

File: components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+          text='Transaction'
+        />
+        <PanelTab
+          isSelected={selectedTab === 'details'}
+          onSubmit={onSelectTab('details')}
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx:163]: Can inline this action

```suggestion
          onSubmit={() => setSelectedTab('details')}
```

File: components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  } = usePendingTransactions()
+  const originInfo = selectedPendingTransaction?.originInfo ?? activeOrigin
+
+  // state
+  const [selectedTab, setSelectedTab] = React.useState<confirmPanelTabs>('transaction')
+
+  // methods
+  const onSelectTab = React.useCallback(
+    (tab: confirmPanelTabs) => () => setSelectedTab(tab),
+  [])
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx:82]: Can remove this method

File: components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx:1]: Lets rename this file to use `snake_case` convention.

`confirm_zcash_transaction_panel.tsx`

File: components/brave_wallet_ui/components/extension/confirm-transaction-panel/zcash-transaction-info.tsx [hunk 9]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/extension/confirm-transaction-panel/zcash-transaction-info.tsx:1]: Lets rename this file to use `snake_case` convention.

`zcash_transaction_info.tsx`

File: components/brave_wallet_ui/components/extension/transaction-box/zcash-transaction-detail-box.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/extension/transaction-box/zcash-transaction-detail-box.tsx:1]: Lets rename this file to use `snake_case` convention.

`zcash_transaction_detail_box.tsx`

**Review by Douglashdaniel — CHANGES_REQUESTED** [review]

**Review by nuo-xu — APPROVED** [review]
iOS++

File: components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx (hunk 6 above)
Comment (cypt4) [reply components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-zcash-transaction-panel.tsx:163]: This is copy of bitcoin transaction panel and will be customized in the future, i'd keep it unchanged to make it easier extracting common parts in the future.

**Review by Douglashdaniel — APPROVED** [review]
++

**Review by supermassive — APPROVED** [review]

**Pull request merged**